src/main/java/com/mongodb/hibernate/translate/attachment/Attachment.java

```diff
@@ -0,0 +1,110 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.hibernate.translate.attachment;
+
+import static java.lang.String.format;
+
+import com.mongodb.hibernate.internal.VisibleForTesting;
+import org.jspecify.annotations.Nullable;
+
+/**
+ * Manages global state while translating SQL AST via {@link org.hibernate.sql.ast.SqlAstWalker}.
+ *
+ * <p>{@code SqlAstWalker} is based on {@code visitor} pattern without returning value from the various visitor methods.
+ * The original goal of {@code SqlAstWalker} is for concatenating SQL string internally; but when translating SQL AST
+ * into MQL, it is highly desirable to return value (e.g. {@link com.mongodb.hibernate.translate.mongoast.AstNode}). To
+ * avoid such dilemma, some attachment/detachment mechanism is implemented as follows:
+ *
+ * <ul>
+ *   <li>when some visitor method needs to return some value, it attaches the value by invoking
+ *       {@link #attach(AttachmentKey, Object)}.
+ *   <li>when some coordinating detacher needs to access the value attached above, it invokes
+ *       {@link #expect(AttachmentKey, Runnable)}; internally the {@code Runnable} will attach value by the above
+ *       attacher.
+ *   <li>both attacher and detacher should provide identical {@link AttachmentKey} to ensure data transferring safety.
+ * </ul>
+ *
+ * @see org.hibernate.sql.ast.SqlAstWalker
+ * @see AttachmentKey
+ */
+@SuppressWarnings("unchecked")
+public final class Attachment {
+
+    private @Nullable AttachmentKey<?> attachmentKey;
+    private @Nullable Object attachedValue;
+
+    @Override
+    public String toString() {
+        return "Attachment{" + "attachmentKey=" + attachmentKey + ", attachedValue=" + attachedValue + "}";
+    }
+
+    @VisibleForTesting(otherwise = VisibleForTesting.AccessModifier.PRIVATE)
+    boolean isBlank() {
+        return this.attachmentKey == null && this.attachedValue == null;
+    }
+
+    /**
+     * Detaches the attached value by invoking some visitor method which returns {@code void} but internally the method
+     * is expected to attach some value by invoking the other {@link #attach(AttachmentKey, Object)} method.
+     *
+     * @param attachmentKey type safe attachment type enum; detacher and attacher should share the identical
+     *     {@code AttachmentKey}; never null
+     * @param attacher some {@link Runnable} wrapper of visitor method which is supposed to attach some value
+     *     internally; never {@code null}
+     * @return the value attached by the attacher; after this method returns, the previous state will be restored; never
+     *     {@code null}
+     * @param <T> the expected Java generic type
+     */
+    public <T> T expect(AttachmentKey<T> attachmentKey, Runnable attacher) {
```

Comment:
It seems to me that neither the method name nor the parameter names here express what they are. This method executes a `Runnable` action, and allows that action to produce a value, which is returned by this method. None of this is expressed by the `expect` name of the method, and `attacher` does not express that it's the action we are executing.

Reply:
the naming was by Jeff in POC project. It is challenging to come up with perfect naming, tbh. In tech doc you mentioned `ReturnType` in lieu of `AttachmentKey`.

@jyemin , I think his naming seems a little bit better